research-app/src/App.vue

```diff
@@ -427,12 +427,54 @@ class AnnotationMessageHandler {
   }
 }
 
+/** This simple class encapsulates how we handle key bindings */
+class KeyboardControlSettings {
+  zoomIn: string;
+  zoomOut: string;
+  moveUp: string;
+  moveDown: string;
+  moveLeft: string;
+  moveRight: string;
+  tiltCW: string;
+  tiltCCW: string;
+  moveAmount: number;
+  tiltAmount: number;
+
+  constructor({
+    zoomIn = "ArrowUp",
+    zoomOut = "ArrowDown",
+    moveUp = "w",
+    moveDown = "s",
+    moveLeft = "d",
+    moveRight = "a",
+    tiltCW = "ArrowLeft",
+    tiltCCW = "ArrowRight",
+    moveAmount = 20,
+    tiltAmount = 20
```

Comment:
This is good but these keybindings are different than the ones that pywwt has been using. We should make these the same as the ones defined here:

https://github.com/WorldWideTelescope/pywwt/blob/master/pywwt/nbextension/static/wwt.html#L122-L187

If I'm remembering the numeric keycodes correctly, the definitions are:

- Zoom in/out are z/x or pgup/pgdn (my parity might be wrong here)
- Directional navigation is i/j/k/l or arrow keys
- Tilt is directional with Alt held down
  - I feel so out-of-touch — do Macs have Alt keys these days? If not we might need to provide an alternative
- There is a built-in delay effect that's lengthened by holding down Shift — I'm not sure what that accomplishes

Reply:
Makes sense - I'll change the keybindings to match pywwt.

My Macbook (Air 2015) doesn't have an Alt key - just Fn, Control, Option, and Command. The Mac keyboards on the Apple store seem to have the same layout, so it doesn't seem to have changed much. An alternative is a good idea, but I'm not sure which makes the most sense - maybe option?

Do we want the Shift delay to be implemented?

Reply:
From playing around with pywwt, I can't figure out any way that the Shift delay helps — I think we can just ignore that.

As for modifiers, there are also complications about which modifiers are intercepted by browsers and/or the OS, and which aren't. A lot of websites seem to use Ctrl-Alt? I bet there's a good explanation somewhere out there about what modifiers you can safely hope to use in the web context ...

Reply:
Ok, sounds good on the Shift delay - one less this to worry about. I'll take a look into modifier keys. Hopefully handling these isn't too painful.

I was thinking a bit more about the KeyboardControlSettings structure, and if we want to have multi-key controls (i.e. Alt + letter), together with the possibility to have multiple key options for each action, I think I'll need to juice it up a little. I have an idea on how to do that which might not be too messy.